node/components.go

```diff
@@ -69,6 +70,7 @@ func baseComponents(cfg *Config, store Store) fx.Option {
 		fx.Provide(services.HeaderService),
 		fx.Provide(services.HeaderStore),
 		fx.Invoke(services.HeaderStoreInit(&cfg.Services)),
+		fxutil.ProvideAs(services.FraudService, new(fraud.Service), new(fraud.Subscriber)),
```

Comment:
do you still need the two return values then from the constructor if you're doing this too?

Reply:
Yes, Daser requires `fraud.Service` and Syncer + State/Service require` fraud.Subscriber`

Reply:
Daser could now require a broadcaster only.

Reply:
btw

Reply:
My question is, then why do you do the `ProvideAs` inside baseComponents?

Reply:
Because different services require different interfaces that are based on `FraudService` and we should explicitly provide them to fx.

Reply:
```diff
diff --git a/node/components.go b/node/components.go
index b7f185d..484febf 100644
--- a/node/components.go
+++ b/node/components.go
@@ -71,7 +71,8 @@ func baseComponents(cfg *Config, store Store) fx.Option {
 		fx.Provide(services.HeaderStore),
 		fx.Invoke(services.HeaderStoreInit(&cfg.Services)),
 		fx.Provide(services.HeaderSyncer),
-		fxutil.ProvideAs(services.P2PSubscriber, new(header.Broadcaster), new(header.Subscriber)),
+		fxutil.ProvideAs(services.P2PSubscriber, new(header.Broadcaster)),
+		fxutil.ProvideAs(services.P2PSubscriber, new(header.Subscriber)),
 		fx.Provide(services.HeaderP2PExchangeServer),
 		// p2p components
 		fx.Invoke(invokeWatchdog(store.Path())),
diff --git a/node/services/service.go b/node/services/service.go
index a65159f..3b0ac2c 100644
--- a/node/services/service.go
+++ b/node/services/service.go
@@ -38,13 +38,13 @@ func HeaderSyncer(
 }
 
 // P2PSubscriber creates a new p2p.Subscriber.
-func P2PSubscriber(lc fx.Lifecycle, sub *pubsub.PubSub) (*p2p.Subscriber, *p2p.Subscriber) {
+func P2PSubscriber(lc fx.Lifecycle, sub *pubsub.PubSub) *p2p.Subscriber {
 	p2pSub := p2p.NewSubscriber(sub)
 	lc.Append(fx.Hook{
 		OnStart: p2pSub.Start,
 		OnStop:  p2pSub.Stop,
 	})
-	return p2pSub, p2pSub
+	return p2pSub
 }
 
 // HeaderService creates a new header.Service.
```

Reply:
May be. @Wondertan wdyt?

Reply:
@renaynay, the problem with the above approach is that it will create multiple instances of the same structure that is supposed to be a singleton. In other words, there will be two FraudServices and two P2PSubscribers hidden via different types.

I agree thought that this is not the prettiest approach, but lets not bikeshed into it very deep.